tfjs-layers/src/layers/nlp/samplers.ts

```diff
@@ -0,0 +1,226 @@
+   *     logits of next token, `hiddenStates` being the representation of
+   *     the next token, and `cache` for next iteration.
+   * @param prompt A 2D integer tensor with shape `[batchSize, maxLength]`. This
+   *     tensor will be iteratively updated column by column with new sampled
+   *     values, starting at `index`.
+   * @param cache A tensor or nested structure of tensors that will be
+   *     updated by each call to `next`. This can be used to cache
+   *     computations from early iterations of the generative loop.
+   * @param index The first index of `prompt` to start sampling at.
+   *     Usually this is set as the length of the shortest non-padded
+   *     sequence in `prompt`.
+   * @param mask A 2D integer tensor with the same shape as `prompt`.
+   *     Locations which are `true` in the mask are never updated during
+   *     sampling. Usually used to mark all locations in the dense prompt
+   *     tensor which were present in a user input.
+   * @param endTokenId The token marking the end of the sequence. If
+   *     specified, sampling will stop as soon as all sequences in the prompt
+   *     produce a `endTokenId` in a location where `mask` is `false`.
+   */
+  apply(
+    next: NextFn,
+    prompt: Tensor,
+    cache?: Tensor,
+    index?: number,
+    mask?: Tensor,
+    endTokenId?: number,
+    hiddenStates?: Tensor
+  ): Tensor {
+    const maxLength = prompt.shape[prompt.shape.length - 1];
+    index = index ?? 0;
+    mask = mask == null ? zerosLike(prompt).cast('bool') : mask.cast('bool');
+    cache = cache ?? tensor([]);
+
+    function cond(prompt: Tensor, index: number): boolean {
+      if (index >= maxLength) {
+        return false;
+      }
+      if (endTokenId == null) {
+        return true;
+      }
+      // Stop if all sequences have produced a *new* endTokenId.
+      const endTokens = prompt.equal(endTokenId).logicalAnd(mask.logicalNot());
+      const promptDone = endTokens.any(-1).cast('bool');
+      return promptDone.all().logicalNot().arraySync() as number === 1;
+    }
+
+    let iter = 0;
+    let logits: Tensor;
+    const maxIterations = maxLength - index;
+    while (iter <= maxIterations && cond(prompt, index)) {
+      // Compute the softmax distribution for the next token.
+      [logits, hiddenStates, cache] = next(prompt, cache, index);
+      const probabilities = softmax(logits.div(this.temperature));
+      // Compute next token.
+      let nextToken = this.getNextToken(probabilities);
+      // Don't overwrite anywhere mask is True.
+      nextToken = nextToken.cast(prompt.dtype);
+      nextToken = where(
+        mask.gather(index, 1),
+        prompt.gather(index, 1),
+        nextToken,
+      );
+      // Update the prompt with the next token.
+      nextToken = nextToken.expandDims(-1);
+      prompt = sliceUpdate(prompt, [0, index], nextToken);
+
+      index += 1;
+      iter += 1;
+    }
+    return prompt;
```

Comment:
This needs to use tf.tidy inside the while loop. Otherwise, it keeps every intermediate tensor, which in gpt2's case is 1024 loops of tensors.

Reply:
Here's a somewhat hacky version I used. It's probably better to use local variables in the tidy and return them from the tidy function. Then, assign the outside variables to the returned ones.

https://github.com/mattsoulanille/tfjs/blob/gpt2_test2/tfjs-layers/src/layers/nlp/samplers.ts#L115-L142

Reply:
got it! For learning purposes, why do we have to manually dispose the oldCache/oldPrompt? I was under the impression the tidy would take care of it. Thanks!